Assign clusters to SCPCL000822 and SCPCL000824 (#683)
AlexsLemonade/OpenScPCA-analysis: analyses/cell-type-ewings/exploratory_analysis/05-cluster-exploration.Rmd

We then look at marker gene expression across the clusters and assign each cluster to a cell type. 

## Setup

```{r packages}
suppressPackageStartupMessages({
  # load required packages
  library(SingleCellExperiment)
  library(ggplot2)
})

# Set default ggplot theme
theme_set(
  theme_bw()
)

# set seed
set.seed(2024)
```

```{r base paths}
# The base path for the OpenScPCA repository, found by its (hidden) .git directory
repository_base <- rprojroot::find_root(rprojroot::is_git_root)

# The current data directory, found within the repository base directory
data_dir <- file.path(repository_base, "data", "current", "SCPCP000015")

# The path to this module
module_base <- file.path(repository_base, "analyses", "cell-type-ewings")

# path to marker genes file
marker_genes_file <- file.path(module_base, "references", "visser-all-marker-genes.tsv")
```

```{r}
# source in helper functions for make_jaccard_matrix() and jaccard()
jaccard_functions <- file.path(module_base, "scripts", "utils", "jaccard-functions.R")
source(jaccard_functions)

# source in helper functions: plot_gene_heatmap() and plot_cnv_heatmap()
# create_classification_df() and create_marker_gene_df()

> validation_functions <- file.path(module_base, "scripts", "utils", "tumor-validation-helpers.R")
> source(validation_functions)

Reviewer: It doesn't look like these functions are used? 

```suggestion
```

```

## Define functions 

```{r, assign_clusters_functions}

> # get clusters for a specified value of k

Reviewer: Expand this "docs" a bit here that it's louvain w/ jaccard since `k` means many different things in clustering land!

cluster_sweep <- function(sce) {
  # first perform clustering across parameters
  cluster_results <- bluster::clusterSweep(reducedDim(sce, "PCA"),
    bluster::NNGraphParam(),
    k = as.integer(seq(5, 40, 5)),
    cluster.fun = "louvain",
    type = "jaccard"
  )

  # turn results into a data frame
  cluster_df <- cluster_results$clusters |>
    as.data.frame() |>
    # add barcode column
    dplyr::mutate(barcodes = colnames(sce)) |>
    # combine all cluster results into one column
    tidyr::pivot_longer(
      cols = ends_with("jaccard"),
      names_to = "params",
      values_to = "cluster"
    ) |>
    # separate out parameters, nn, function, and type into their own columns
    dplyr::mutate(
      nn_param = stringr::word(params, 1, sep = "_") |>
        stringr::str_replace("k.", "k_"),
      cluster_fun = stringr::word(params, 2, sep = "_") |>
        stringr::str_remove("cluster.fun."),
      cluster_type = stringr::word(params, -1, sep = "_") |>
        stringr::str_remove("type.")
    ) |>
    # remove combined params column
    dplyr::select(-params)

  return(cluster_df)
}
```

```{r, cluster_stats_functions}

> # cluster statistics functions

Reviewer: ```suggestion
# cluster statistics functions

```

        as.data.frame() |>
        tibble::rownames_to_column("barcodes")

      # join into one data frame to return
      stats_df <- sil_df |>
        dplyr::left_join(purity_df, by = "barcodes")

      return(stats_df)
    }) |>
    dplyr::bind_rows(.id = "nn_param")

  return(all_stats_df)
}

# calculate cluster stability for a single set of clusters using ari
# bootstrap and get ari for clusters compared to sampled clusters
# re-clusters and gets ari across 20 iterations
get_ari <- function(pcs,
                    clusters,
                    k) {
  ari <- c()
  for (iter in 1:20) {
    # sample cells with replacement
    sample_cells <- sample(nrow(pcs), nrow(pcs), replace = TRUE)
    resampled_pca <- pcs[sample_cells, , drop = FALSE]

    # perform clustering on sampled cells
    resampled_clusters <- get_clusters(resampled_pca, k)

    # calculate ARI between new clustering and original clustering
    ari[iter] <- pdfCluster::adj.rand.index(resampled_clusters, clusters[sample_cells])
  }

  ari_df <- data.frame(
    ari = ari,
    k_value = k
  )
}

# get cluster stability for each nn_param cluster results are available for

> # calculated the ari between original clusters and shuffled clusters 20 times

Reviewer: this was left over from copy/paste probably 
```suggestion
```

classification_df <- classification_df |>
  dplyr::left_join(gene_exp_df, by = c("barcodes", "library_id"))
```

## Clustering

Below we perform Louvain, Jaccard clustering, varying `k`. 
The minimum `k` is 5 and the maximum `k` is 40 with a step size of 5. 

```{r}
# perform clustering across k = 5-40 with increments of 5
all_cluster_results <- sce_list |>
  purrr::map(cluster_sweep) |>
  dplyr::bind_rows(.id = "library_id")
```

```{r}
# get umap embeddings and combine into a data frame with cluster assignments
umap_df <- sce_list |>
  purrr::map(\(sce){
    df <- sce |>
      scuttle::makePerCellDF(use.dimred = "UMAP") |>
      # replace UMAP.1 with UMAP1 and get rid of excess columns
      dplyr::select(barcodes, UMAP1 = UMAP.1, UMAP2 = UMAP.2)
  }) |>
  dplyr::bind_rows(.id = "library_id") |>
  dplyr::left_join(all_cluster_results, by = c("library_id", "barcodes"))
```

Below we visualize the cluster assignments using each parameter on a UMAP. 
This isn't particularly useful because of clusters, but it can be helpful to see any values of K that may have obvious over or under clustering. 

```{r, fig.height=15, fig.width=10}

> # look at clustering results across all params

Reviewer: For this plot you're going to need to tweak the factors, since `k_5` is last because the computer thinks 5>40 😞

Reviewer: Actually also, this plot is really hard to take in because it's so long. Why don't we just plot the two datasets separately so each has its own `facet_wrap`? We aren't comparing dataset UMAPs directly to each other anyways and a shared legend isn't particularly meaningful.

  ) +
  theme(aspect.ratio = 1)
```

### Clustering statistics 

Below we calculate a series of statistics: 

- Average silhouette width: This metric evaluates cluster separation. 
Cells with large positive silhouette widths are closer to other cells in the same cluster than to cells in different clusters. 
- Average cluster purity: This metric also evaluates cluster separation and tells us the proportion of neighboring cells that are assigned to the same cluster. 
Higher purity values indicate that clusters are well separated. 
- Cluster stability: This evaluates how stable the clustering is to input data. 
Higher values of cluster stability indicate more reproducible clusters. 

```{r, warning=FALSE}
# get a combined stats dataframe with purity and width for all clusters
all_stats_df <- sce_list |>
  purrr::imap(\(sce, library_id){
    get_cluster_stats(sce, library_id, all_cluster_results)
  }) |>
  dplyr::bind_rows(.id = "library_id")
```

```{r}
# summarize the silhouette width and purity for each nn_param used
summary_stats_df <- all_stats_df |>
  dplyr::group_by(library_id, nn_param) |>
  dplyr::summarise(
    mean_width = mean(width),
    mean_purity = mean(purity)

>   ) |>

Reviewer: I'd like to see some error lines in these plots too to assess overlap, if we're plotting means (that said, did you try maybe a `sina` with a red mean dot like the downstream analysis clustering report?). On one hand I'd say toss an a standard error calculation here too, but I actually you can probably skip a lot (all?) of this wrangling using `stat_summary()` which should plot your mean & SE automagically.

```

```{r}
# silhouette width for different params
ggplot(summary_stats_df, aes(x = nn_param, y = mean_width)) +
  geom_point() +
  facet_wrap(vars(library_id),
    scales = "free_y"
  ) +
  labs(y = "Average silhouette width")
```

```{r}
# cluster purity for different params
ggplot(summary_stats_df, aes(x = nn_param, y = mean_purity)) +
  geom_point() +
  facet_wrap(vars(library_id),
    scales = "free_y"
  ) +
  labs(y = "Average cluster purity")
```

Here we see that both silhouette width and cluster purity peak around k=20 or k=25. 
For `SCPCL000822`, increasing k after 25 does not change the width or purity, 
For `SCPCL000824`, increasing k after 20-25 decreases the width and keeps purity around the same. 

```{r, warning=FALSE, message=FALSE}
# calculate cluster stability
stability_df <- sce_list |>
  purrr::imap(\(sce, id){
    get_cluster_stability(sce, id, all_cluster_results)
  }) |>
  dplyr::bind_rows(.id = "library_id") |>
  dplyr::mutate(
    k_value = as.factor(k_value),
    # make sure that k = 5 comes first
    k_value = forcats::fct_relevel(k_value, "5", after = 0)
  )

# plot stability across all values of k
ggplot(stability_df, aes(x = k_value, y = ari)) +

>   geom_point() +

Reviewer: just so we can make sure we're seeing the points, and I suggested my personal favorite width parameter

```suggestion
  geom_jitter(width = 0.1) +
```

```{r}
cluster_classification_df <- all_cluster_results |>
  # add in classifications from singler
  dplyr::left_join(classification_df, by = c("library_id", "barcodes")) |>
  dplyr::mutate(
    # first grab anything that is tumor and label it tumor
    # NA should be unknown
    singler_annotation = dplyr::case_when(
      stringr::str_detect(singler_annotation, "tumor") ~ "tumor",
      is.na(singler_annotation) ~ "unknown", # make sure to separate out unknown labels
      .default = singler_annotation
    ) |>
      forcats::fct_relevel("tumor", after = 0),
    # get the top cell types for plotting later
    singler_lumped = singler_annotation |>
      forcats::fct_lump_n(7, other_level = "All remaining cell types", ties.method = "first") |>
      forcats::fct_infreq() |>
      forcats::fct_relevel("All remaining cell types", after = Inf)
  )
```

```{r, fig.height=10, fig.width=7}
# get heatmap for each library showing cluster assignments vs. singler assignments
names(sce_list) |>
  purrr::map(\(id){
    cluster_celltype_heatmap(
      cluster_classification_df,

>       id

Reviewer: need to set factors for this plot, 5 is last again.

Reviewer: I wonder if we should turn off (at least y-axis) clustering for these heatmaps so that the cell labels are always in the same order. I think that will make it easier to interpret.

    )
  })
```

From these plots, it's obvious that K = 5 is over clustering and cell types are spread across a lot of clusters. 

> However, for the other values of K, there doesn't seem to be any striking difference that are obvious by eye.  

Reviewer: ```suggestion
However, for the other values of K, there don't seem to be any striking differences that are obvious by eye.  
```

- Any clusters with high immune marker gene expression also correspond to clusters that are predominantly "macrophage" in the heatmaps, so they will be labeled as "macrophage". 
- Any clusters with high endothelial marker gene expression also correspond to clusters that are predominantly "endothelial cell" in the heatmaps, so they will be labeled as "macrophage". 
- There is one cluster in `SCPCL000822` that has high mesenchymal gene expression and is predominantly chondrocytes so we will label this cell as "chondrocytes". 
- All other clusters with high mesenchymal gene expression appear to have a mix of chondrocytes, fibroblasts, and smooth muscle cells so we will just label these as "mesenchymal-like cells". 

```{r}
# get umap embeddings for plotting later
umap_df <- umap_df |>
  dplyr::select(library_id, barcodes, UMAP1, UMAP2) |>
  dplyr::distinct()

# assign clusters based on marker gene expression and cell types that are prominent in that cluster
# use the jaccard matrices to help guide assignments
cluster_umap_df <- cluster_classification_df |>
  dplyr::left_join(umap_df, by = c("library_id", "barcodes")) |>
  dplyr::filter(nn_param == "k_25") |>

>   dplyr::mutate(

Reviewer: Somewhere it is worth noting (comments? bullets? you decide!) which clusters are _not_ being assigned. Looks like 10&11 for `SCPCL000822` and 9 for `SCPCL000824`.

      # cluster 3 has predominantly chondrocytes and has MSC marker gene expression
      library_id == "SCPCL000822" & cluster == 3 ~ "chondrocyte",
      # clusters with high "immune" marker gene expression are macrophage
      library_id == "SCPCL000822" & cluster == 6 ~ "macrophage",
      # cluster 8 has high MSC marker gene expression and a mix of MSC like cell types
      library_id == "SCPCL000822" & cluster == 8 ~ "mesenchymal-like cell",
      library_id == "SCPCL000824" & cluster %in% c(1, 2, 3, 4, 5, 6, 11) ~ "tumor",
      library_id == "SCPCL000824" & cluster == 8 ~ "endothelial cell",
      # cluster 10 has high immune markers and is mostly macrohpages
      library_id == "SCPCL000824" & cluster == 10 ~ "macrophage",
      # cluster 7 has high MSC marker gene expression and a mix of MSC like cell types
      library_id == "SCPCL000824" & cluster == 7 ~ "mesenchymal-like cell",
      # if they didn't have high expression of any markers, we label them as unknown
      .default = "unknown"
    ),
    # make tumor first
    cluster_celltype = forcats::fct_relevel(cluster_celltype, "tumor", after = 0)
  )
```

> ```{r}

Reviewer: I'd make this plot bigger, maybe move legend to bottom as needed for spacing. Can't directly suggest changes to the code chunk though since too many backticks...

# look at cell type assignments on the UMAP
ggplot(cluster_umap_df, aes(x = UMAP1, y = UMAP2, color = cluster_celltype)) +
  geom_point(alpha = 0.5, size = 0.5) +
  facet_grid(cols = vars(library_id)) +
  theme(aspect.ratio = 1) +
  guides(color = guide_legend(override.aes = list(alpha = 1, size = 1.5))) +
  scale_color_brewer(palette = "Dark2")
```

> We can also validate that these cells do in fact have higher expression of the marker genes as compared to other cells by plotting the marker gene expression on UMAPs. 

Reviewer: It's hard to interpret these plots since they paired by cell type, not by sample. Is it possible to instead have two plots - one for each sample and faceted by cell type? (right now it's 3 plots, one for each cell type but faceted by sample)

```{r}
# grab columns that contain marker gene sums
marker_gene_columns <- colnames(cluster_umap_df)[which(endsWith(colnames(cluster_umap_df), "_sum"))]

> # create individual density plots and combine into one

Reviewer: ```suggestion
# create individual umaps and combine into one
```

  })
```

### Clustering statistics 

Below we calculate a series of statistics: 

- Average silhouette width: This metric evaluates cluster separation. 
Cells with large positive silhouette widths are closer to other cells in the same cluster than to cells in different clusters. 
- Average cluster purity: This metric also evaluates cluster separation and tells us the proportion of neighboring cells that are assigned to the same cluster. 
Higher purity values indicate that clusters are well separated. 
- Cluster stability: This evaluates how stable the clustering is to input data. 
Higher values of cluster stability indicate more reproducible clusters. 

```{r, warning=FALSE}
# get a combined stats dataframe with purity and width for all clusters
all_stats_df <- sce_list |>
  purrr::imap(\(sce, library_id){
    get_cluster_stats(sce, library_id, all_cluster_results)
  }) |>
  dplyr::bind_rows(.id = "library_id") |>
  dplyr::mutate(
    # make sure the order is correct
    nn_param = forcats::fct_relevel(nn_param, "k_5", after = 0)
  )
```

```{r}
# silhouette width for different params
library_ids |>
  purrr::map(\(id){

>     plot_cluster_stats(all_stats_df, width, id)

Reviewer: Can we do a ....

```suggestion
    plot_cluster_stats(all_stats_df, width, id) + ggtitle(id)
```

  })
```

Here we see that both silhouette width and cluster purity peak around k=20 or k=25. 
For `SCPCL000822`, increasing k after 25 does not change the width or purity, 
For `SCPCL000824`, increasing k after 20-25 decreases the width and keeps purity around the same. 

```{r, warning=FALSE, message=FALSE}
# calculate cluster stability
stability_df <- sce_list |>
  purrr::imap(\(sce, id){
    get_cluster_stability(sce, id, all_cluster_results)
  }) |>
  dplyr::bind_rows(.id = "library_id") |>
  dplyr::mutate(
    k_value = as.factor(k_value),
    # make sure that k = 5 comes first
    k_value = forcats::fct_relevel(k_value, "5", after = 0)
  )

# plot stability across all values of k
ggplot(stability_df, aes(x = k_value, y = ari)) +
  geom_jitter(width = 0.1) +
  facet_grid(rows = vars(library_id)) +
  labs(title = "Cluster stability") +
  stat_summary(
    aes(group = k_value),
    color = "red",
    # median and quartiles for point range
    fun = "median",
    fun.min = function(x) {
      quantile(x, 0.25)
    },
    fun.max = function(x) {
      quantile(x, 0.75)
    }
  )
```

Here we see that cluster stability increases as k increases for both libraries. 

> We see a plateau start to appear around k = 20 and k = 25. 

Reviewer: Looking at this again, I actually now wonder if we should be at k=30/35 for `SCPCL000824` with this metric? What do you think?

  dplyr::left_join(classification_df, by = c("library_id", "barcodes")) |>
  dplyr::mutate(
    # first grab anything that is tumor and label it tumor
    # NA should be unknown
    singler_annotation = dplyr::case_when(
      stringr::str_detect(singler_annotation, "tumor") ~ "tumor",
      is.na(singler_annotation) ~ "unknown", # make sure to separate out unknown labels
      .default = singler_annotation
    ) |>
      forcats::fct_relevel("tumor", after = 0),
    # get the top cell types for plotting later
    singler_lumped = singler_annotation |>
      forcats::fct_lump_n(7, other_level = "All remaining cell types", ties.method = "first") |>
      forcats::fct_infreq() |>
      forcats::fct_relevel("All remaining cell types", after = Inf),
    nn_param = forcats::fct_relevel(nn_param, "k_5", after = 0)
  )
```

```{r, fig.height=10, fig.width=7}
# get heatmap for each library showing cluster assignments vs. singler assignments
library_ids |>
  purrr::map(\(id){
    cluster_celltype_heatmap(
      cluster_classification_df,
      id
    )
  })
```

From these plots, it's obvious that K = 5 is over clustering and cell types are spread across a lot of clusters. 

> However, for the other values of K, there don't seem to be any striking differences that are obvious by eye.  

Reviewer: im intrigued by the macrophage/cluster 9 interaction for `SCPCL000824`, as well as some of what's going on with cluster 10. Looks like things are bouncing back and forth from k=20 -> k=40 for a decent number of the cells?

Author: I think the exact numbers of the clusters here can't really be compared between multiple parameters. We don't actually know that the cells that were in cluster 9 when we cluster with one value of k are the same cells that are in cluster 9 or 10 in another value of k. What I see is there is a cluster that has both macrophages and some cells from the "All remaining cell types" class. With k=25, this is cluster 10 and with k=30 this is cluster 9. 

That being said I do agree that maybe k=30 is a little bit better for SCPCL000824. Mostly because of the increase in purity and increase in stability that is shown between K=25 to 30.

Reviewer: > I think the exact numbers of the clusters here can't really be compared between multiple parameters. We don't actually know that the cells that were in cluster 9 when we cluster with one value of k are the same cells that are in cluster 9 or 10 in another value of k.

Haha of course.. Friday afternoon brain 😄

Reviewer: > That being said I do agree that maybe k=30 is a little bit better for SCPCL000824. Mostly because of the increase in purity and increase in stability that is shown between K=25 to 30.

Do you plan to run these libraries again through the template notebook? If so, I'd leave this one is, and just indicate in the text that 30 could be good too, and save changes for later. Otherwise if this is it for these libraries, I'd just go ahead and use 30.

  geom_point(alpha = 0.5, size = 0.5) +
  facet_grid(cols = vars(library_id)) +
  theme(
    aspect.ratio = 1,
    legend.position = "bottom"
  ) +
  guides(color = guide_legend(override.aes = list(alpha = 1, size = 1.5))) +
  scale_color_brewer(palette = "Dark2")
```

We can also validate that these cells do in fact have higher expression of the marker genes as compared to other cells by plotting the marker gene expression on UMAPs. 

```{r, fig.height=7, fig.width=7}
# grab columns that contain marker gene sums
marker_gene_columns <- colnames(cluster_umap_df)[which(endsWith(colnames(cluster_umap_df), "_sum"))]

# for each sample show a faceted umap by cell type
# color is gene expression of all marker genes for that cell type
library_ids |>
  purrr::map(\(id){
    plot_df <- cluster_umap_df |>
      dplyr::filter(library_id == id) |>
      tidyr::pivot_longer(
        cols = all_of(marker_gene_columns),
        names_to = "cell_type",
        values_to = "gene_expression"
      ) |>
      dplyr::mutate(cell_type = stringr::str_remove(cell_type, "_sum"))

    ggplot(plot_df, aes(x = UMAP1, y = UMAP2, color = gene_expression)) +
      geom_point(alpha = 0.5, size = 0.2) +

>       facet_wrap(vars(cell_type)) +

Reviewer: I wonder if `scales="free"` would help with the color scale, but also it might just do something bonkers, or do nothing at all!...never tried it before for such a circumstance..
